### ISSUE TITLE
Allow compiler wrapper to prepend-path

### ISSUE THREAD

**Issue by skosukhin** [spack/spack#5019]
Does it look like a good idea to extend #2275 by introducing 'prepend-path' keyword for compilers.yaml (yes, let's just integrate the whole lmod =)?

For example, that could help to tell Intel compiler to use particular binutils. Unfortunately, just setting environment variables is not enough for this: https://software.intel.com/en-us/comment/1910316

**Comment by skosukhin** [comment]
@adamjstewart @alalazo @scheibelp @tgamblin could anyone, please, explain to me why we need SPACK_ENV_SET_* and SPACK_ENV_TO_SET variables? It looks like we set an environment variable from compilers.yaml twice: first, in `set_build_environment_variables()` and then again in the compiler wrapper. Why do we do it in the wrapper? Thank you.

**Comment by alalazo** [comment]
@skosukhin I am reading this just now. Definitely +1 to use the keywords that are used for module files (`set`, `unset`, `append-path`, `prepend-path`). 

> could anyone, please, explain to me why we need SPACK_ENV_SET_* and SPACK_ENV_TO_SET variables?

I tried to reconstruct how things got added. I need to think more about that, but the reason those variables are there may be vestigial. The variables you are referring to were added in #2275 which, as @scheibelp noted, was copied from #943 that dates 05/2016. 

At that time the compiler wrapper was also unsetting `LD_LIBRARY_PATH` and similar variables and thus a mechanism like `SPACK_ENV_SET_*` was the only way to work around this mechanism in case one of those variables was needed for the compiler.

In #2074 (that dates 10/2016, so 5 months later) the lines unsetting `LD_LIBRARY_PATH` etc. within the copmiler wrapper were removed. Imo it's very likely that `SPACK_ENV_SET_*` and friends are not needed anymore.

**Comment by skosukhin** [comment]
I have a question about the order of modifications. Let's assume we don't set `SPACK_ENV_SET_*` and `SPACK_ENV_TO_SET` and don't introduce something like `SPACK_ENV_PREPEND_PATH_*`.

Then we have the following sources of modification of `PATH` (the order of modifications is taken from the code):
1. Spack prepends `PATHs` of dependencies.
2. Platform can modify `PATH` by overloading method `setup_platform_environment()`.
3. Package can modify `PATH` by overloading method `setup_dependent_environment()`.
4. Package can modify `PATH` by overloading method `setup_environment()`.
5. Compiler can modify `PATH` with the modules specified in its configuration `modules`.
6. Compiler will be able (if the discussed feature is implemented) to modify `PATH` with the commands specified in its configuration `environment` (that's where I am going to put it).
7. Target can modify `PATH` by loading its `module_name`.
8. `PATH` can be modified by external modules.

Do you agree with this order? What if we specify path to `binutils` in the `environment` section of the compiler configuration, and `binutils` is also among the dependencies? Which one should be first in the path? What if `binutils` is an external module? What should happen in this case?

**Comment by svenevs** [comment]
If the compiler can set environment variables, shouldn't those be first?

**Comment by alalazo** [comment]
@skosukhin Should we to put these two:

1. Package can modify PATH by overloading method setup_dependent_environment().
1. Package can modify PATH by overloading method setup_environment().

as last? I am wondering this because they are basically the finest line of defense against issues in the environment, as they can act on a package-by-package base.

As an example, say some environment variable set in the compiler, target or in some external module (that are now evaluated after packages) conflicts in a pathological way with a package build. With the order above the package can't do anything to prevent this, and won't build. Putting package modifications last we reserve ourselves the chance to isolate that case and act on it, e.g.:
```
if '%compiler' in self.spec:
     spack_env.unset('PROBLEMATIC_VARIABLE')
```
or something similar. Does it make sense?

**Comment by becker33** [comment]
Closed by #8996
